server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MirroringMigrationService.java

```diff
@@ -0,0 +1,276 @@
+import com.linecorp.centraldogma.server.mirror.MirrorCredential;
+import com.linecorp.centraldogma.server.storage.project.Project;
+import com.linecorp.centraldogma.server.storage.project.ProjectManager;
+
+class MirroringMigrationService {
+
+    private static final Logger logger = LoggerFactory.getLogger(MirroringMigrationService.class);
+
+    @VisibleForTesting
+    static final String PATH_LEGACY_MIRRORS = "/mirrors.json";
+    @VisibleForTesting
+    static final String PATH_LEGACY_CREDENTIALS = "/credentials.json";
+    private static final String PATH_LEGACY_MIRRORS_BACKUP = PATH_LEGACY_MIRRORS + ".bak";
+    private static final String PATH_LEGACY_CREDENTIALS_BACKUP = PATH_LEGACY_CREDENTIALS + ".bak";
+
+    private final ProjectManager projectManager;
+
+    @Nullable
+    private List<String> shortWords;
+
+    MirroringMigrationService(ProjectManager projectManager) {
+        this.projectManager = projectManager;
+    }
+
+    void migrate() throws Exception {
+        for (Project project : projectManager.list().values()) {
+            migrateCredentials(project);
+            // Update the credential IDs in the mirrors.json file.
+            migrateMirrors(project);
+        }
+        shortWords = null;
+    }
+
+    private void migrateMirrors(Project project) throws Exception {
+        final List<MirrorCredential> credentials = project.metaRepo().credentials().join();
+        final ArrayNode mirrors = getMetaData(project, PATH_LEGACY_MIRRORS);
+        if (mirrors == null) {
+            return;
+        }
+
+        final Set<String> mirrorIds = new HashSet<>();
+        for (JsonNode mirror : mirrors) {
+            if (!mirror.isObject()) {
+                logger.warn("A mirror config must be an object: " + mirror);
+                continue;
+            }
+            try {
+                migrateMirror(project, (ObjectNode) mirror, mirrorIds, credentials);
+            } catch (Exception e) {
+                logger.warn("Failed to migrate a mirror config: " + mirror, e);
+            }
+        }
+
+        // Back up the old mirrors.json file and don't use it anymore.
+        rename(project, PATH_LEGACY_MIRRORS, PATH_LEGACY_MIRRORS_BACKUP);
+    }
+
+    private void migrateMirror(Project project, ObjectNode mirror, Set<String> mirrorIds,
+                               List<MirrorCredential> credentials) throws Exception {
+        String id;
+        final JsonNode idNode = mirror.get("id");
+        if (idNode == null) {
+            // Fill the 'id' field with a random value if not exists.
+            id = generateRandomIdForMirror(project.name(), mirror);
+        } else {
+            id = idNode.asText();
+        }
+        id = uniquify(id, mirrorIds);
+        mirror.put("id", id);
+
+        fillCredentialId(mirror, credentials);
+        if (mirror.get("schedule") == null) {
+            mirror.put("schedule", DEFAULT_SCHEDULE);
```

Comment:
I hesitated to change the value because, currently, the default schedule is 1 minute if no schedule is set. If we change the default schedule to 5 or 10 minutes, that could impact users.

If necessary, I prefer handling the breaking change in a separate PR to easily know the change from the PR title.

Reply:
Yeah, we need to handle this separately. Let me create an issue for this.